### ISSUE TITLE
Multiple folders for code snippets // external folders?

### ISSUE THREAD

**Issue by jondoering** [elyra-ai/elyra#1133]
Hi, 

I really like the code snippet plugin for Jupyter lab - something I missed for a long time.
However, the current way of storing snippets in individual json files in one folder, makes it hard to share snippets and put them in version control. I use symbolic links to link my code-snippets metadata folder with an external git folder, but this is rather hacky - and it does not allow to have different repositories for different snippets. What would be a killer feature to share code between team members - or teams - in an organization. 
I looked into the code, but the structure of the FileStorage class makes it hard to add something like this.
Are there any plans to improve here - and make snippets more git friendly? 

Thanks, 
Jonathan

**Comment by lresende** [comment]
That's a very interesting spin on the code snippet, from what you are suggesting we need to consider multiple sources and enable storage in git. 

About the git side, how do you see the commits/push happening, as actions triggered by the user? or something more automated associated with the save of a code snippet content?

**Comment by kevin-bates** [comment]
Hi @jondoering - thank you for opening this issue.  I agree, that would be a killer feature. 

(I apologize for the length here.  I got to talking and things just came out.  I think we have a few options, but things are in the early stages here, so, in that sense, this is all good.)

I do have a couple of questions and some thoughts.  First, I just wanted to point out that the current usage is, as you noted, purely flat and file-based such that we can easily locate the metadata "instances".  The idea behind the storage concept is that a different implementation would be used to store "instances" in different kinds of mechanisms: NoSQL, K-V Stores, etc. Git seems like a logical idea.  Each storage area would know how to deal with it's "structure" and may require some additional (storage-wide) properties.  I think it's important to keep in mind that code-snippets are simply a metadata instance and, in that sense, are less flexible since all kinds (namespaces/schemas) of metadata instances adhere to the same principles.

Unfortunately, we currently don't have a way to specify different storage behaviors for different Namespaces.  I've talked in the past about [promoting Namespace from a String to a Class](https://github.com/elyra-ai/elyra/issues/701#issuecomment-716888388), where it would contain a list of schemas (currently, the schemas reference the Namespace (string)), storage characteristics for _its_ instances, and any other schema-and-instance-spanning properties that might be necessary. 

(I just saw Luciano's response while writing this up and had similar questions)

- Do you foresee the entire set of metadata properties being stored in git?  Or just the code snippet code itself?
- Would code snippet instances (or pieces if only storing the code itself) reside in the same repo, be in separate repos, or in the same repo but separate locations/folders?

If the entire instance is to be stored in Git, then it seems like a "GitMetadataStore" class could be implemented and persist the metadata instances into/out of git.  Which org, repo, and creds to use would be attributes of the storage class or be something derived from the metadata instance itself (although I'd prefer the storage class to not have knowledge about the instance data - except for perhaps _system-level_ properties - i.e., those properties not within the `metadata` "stanza"). I think we'd need to introduce the concept of namespace-based storage since not all namespaces would want their instances to be stored in Git.

If only the code itself were to be stored in git, then I could see this being a place we could introduce a [`meta-behavior`](https://github.com/elyra-ai/elyra/issues/518) that handles storage/retrieval of a given property in a different manner.  This particular meta-behavior would have the necessary information to persist/retrieve the code itself from git.

We should also explore the idea of levels within a namespace.  Different storage mechanisms might handle these levels differently.  For example, a "GitMetadataStore" class may see a (Namespace-relative) path of "acme-inc/ai-snippets/spark-session" and know the first two "segments" represent the organization and repository, respectively, with the last being the code-snippet name, etc.  This same "pathing" could also have the meaning of a sub-directory structure to the FileMetadataStore.  Again, the property that contains this "pathing" may be associated with, say, a _location-oriented_ meta-behavior such that this property value is used to convey "location" to the Storage class - and lets the storage class interpret the "pathing".  By using this property as a meta-behavior, we still retain the idea that Storage is essentially agnostic to anything in the `metadata` stanza.

I just realized another area this could have some utility if only saving the code itself in git - by introducing a CodeSnippet metadata class.  Each schema supports a [`metadata_class_name` _`meta-behavior`_](https://github.com/elyra-ai/elyra/blob/master/elyra/metadata/schemas/metadata-test2.json#L7) in which the metadata manager will instantiate it for each instance.  If no class is specified the metadata instance is essentially treated as a JSON blob.  The framework then calls the class's implementations for [`post_load()`, `pre_save()`, and `pre_delete()`](https://github.com/elyra-ai/elyra/blob/master/elyra/metadata/metadata.py#L45-L66) and, the Git implementation could pull the code snippet from git in `post_load()` and seed the code-snippet lines list.  Then, similarly, perform the appropriate operations against git for the `pre_save()` and `pre_delete()` methods.  In using this approach, the metadata class (unlike the storage class) _would_ be privy to metadata instance data and schema.

> Are there any plans to improve here - and make snippets more git friendly?

So, I guess while we didn't necessarily have specific plans about code-snippets and git, we do have "plans" (more in the "idea" stages) about designing things in a way so that different behaviors can be introduced to specific namespaces and instances.  

It would be great to collaborate on this.

**Comment by jondoering** [comment]
Hi, 
Thanks for taking that up and all those already very well thought thru ideas. 

To add my 2 cents on the open questions: 

> Do you foresee the entire set of metadata properties being stored in git? Or just the code snippet code itself?
I do see merrit in having the full metadata property stored, given that additional data such as language and tags are stored as well. It would also be easy to manage since it is a JSON file, I guess. 

> Would code snippet instances (or pieces if only storing the code itself) reside in the same repo, be in separate repos, or in the same repo but separate locations/folders?
IMO a really great implementation, would allow to have multiple repos combined - this would allow to have people open up specific code-snippets / template repos which can be reused and would be combined in the UI as one searchable list. 
E.g. I would start to build use case specific snippet repositories (e.g. for different librarie (keras, tensorflow) or different workflow steps (data preprocessing, modeling)  and could easily share that with my team. 

> About the git side, how do you see the commits/push happening, as actions triggered by the user? or something more automated associated with the save of a code snippet content?
I would not have a problem with this as individual action of a user, in the same way as I commit the changes of my code to a repo. Having auto-commit/push might be even more confusing.

**Comment by kevin-bates** [comment]
>>About the git side, how do you see the commits/push happening, as actions triggered by the user? or something more automated associated with the save of a code snippet content?

>I would not have a problem with this as individual action of a user, in the same way as I commit the changes of my code to a repo. Having auto-commit/push might be even more confusing.

Could someone describe the real differences between "actions triggered by the user" and "something more automated associated with the save of a code snippet content"?  Is there another way to save content other than an action triggered by the user?  I don't understand where the "automated" portion of this conversation comes from?

If the implication is that we'd persist the code snippet metadata differently than the code snippet _code_ itself, this is what I referenced above with meta-behaviors.  However, both actions (the persistence of the metadata properties and the code snippet code) are triggered by the user's (single) action of saving the code snippet.  I suppose the same goes for the pluggable storage area approach - the git integration would not occur until the user says so.

Since the git "workflow" is more of a checkpoint-checkpoint-then-save model (e.g., commit-commit-then-push), it seems we'd need to introduce another "_checkpointing_ action", made by the user, that essentially commits changes, but doesn't push changes - with "save" tantamount to "push".  If that's the model here (rather than the save triggering both a commit-and-push), then I would suggest we not use the metadata framework at all for code snippets, but instead, use the "git" extension to manage code-snippet "files".  That should provide all the fine-grained interactions necessary.

**Comment by lresende** [comment]
> Could someone describe the real differences between "actions triggered by the user" and "something more automated associated with the save of a code snippet content"? Is there another way to save content other than an action triggered by the user? I don't understand where the "automated" portion of this conversation comes from?

While I think it's a good idea to enable git integration to code-snippet, I want to avoid having to re-create another Git UI to enable push/pull/commit/diff/resolve conflict, etc.

A simple way would be to "enable multiple folders" to code-snippet, sort of a federated code snippet concept, but let the user handle the pull/push/resolve conflict manually (e.g. via terminal). I think we could, at most, do a commit while saving, but then we would need to start introducing the concept of title of the change and ask the user on every save, etc

The other approach is to handle the full-blown git integration but in this case, we would probably depart from the generic data-driven ui and even maybe from the metadata layer. 

> If the implication is that we'd persist the code snippet metadata differently than the code snippet code itself, this is what I referenced above with meta-behaviors. However, both actions (the persistence of the metadata properties and the code snippet code) are triggered by the user's (single) action of saving the code snippet. I suppose the same goes for the pluggable storage area approach - the git integration would not occur until the user says so.

I don't think that we are talking about separating different pieces of the code-snippet, but actually enabling a different store, (GitMetadataStore) and adding a new higher level class to aggregate (FederatedMetadataManager). We would also enable the user to configure/add multiple git repositories.


I believe that, if we take the approach described above, we could get something quickly where initially the user would have to handle some tasks manually and then iterate to a better overall user experience.  Similar to how we started with runtime-configuration. 

@kevin-bates @jondoering Thoughts?

@jondoering would you like to help with the implementation by providing a patch?

**Comment by kevin-bates** [comment]
> I don't think that we are talking about separating different pieces of the code-snippet, but actually enabling a different store, (GitMetadataStore) 

This was one of my questions, but, yes, it sounds like the idea would be to store the complete code-snippet in git - which implies a metadata store implementation - I agree.

> and adding a new higher level class to aggregate (FederatedMetadataManager). We would also enable the user to configure/add multiple git repositories.

I'm not following the 'federated' 'aggregated' ideas and how a 'FederatedMetadataManager' comes into play or what it even does. 

How are these folders and git repositories expressed in terms of the schema and how are different code-snippet instances associated with the various folders, etc?

Would different Namespaces have (potentially) different metadata manager implementations?

**Comment by lresende** [comment]
> I'm not following the 'federated' 'aggregated' ideas and how a 'FederatedMetadataManager' comes into play or what it even does.

I was thinking that a store would be associated with a git repo, and a federated higher level class would aggregate the values from multiple git repositories/stores

>How are these folders and git repositories expressed in terms of the schema and how are different code-snippet instances associated with the various folders, etc?

No changes to the schema, it's more like configuring the gitstore to associate it with a git repo

**Comment by lresende** [comment]
@jondoering I discussed this a little with @kevin-bates and we were trying to find a simple initial way to start solving this:

Some initial thoughts:
- Enable hierarchical code-snippets rooted in the existing code-snippet metadata folder
- Each git repository would be cloned as a sub-directory on the root code-snippet metadata folder
- Code snippet would be enhanced to properly update code-snippets from the sub-directories
- New code-snippets would be saved in the root (or we could assume a name with x/name goes to sub-directory x)
- Git operations would be handled manually by the user

In the future, we could look into enhancing the git extension to point it to directories outside of the jupyter working directory to enable git operations with code-snippets. 

@kevin-bates Did I miss anything? 
@jondoering Thoughts?

**Comment by kevin-bates** [comment]
@lresende - thanks for this response.  I had meant to respond sooner with some thoughts of my own and I'll share those below.
> @kevin-bates Did I miss anything?

This essentially covers things although if we're not providing automatic git operations, then the _encoding_ of the meaning of the sub-directories is nothing more than a suggestion (for now).

> we could assume a name with x/name goes to sub-directory x

Just to clarify, _`x`_ could by a set of sub-directories _`x/y/z`_ prefixing `name`.

When thinking about this further after our discussion, I think we should preserve the current usage of `name` as it is today (i.e., not overload meaning into its value), but introduce an optional _transient_ system property to schemas (for now it might only exist in the code-snippets schema but the following behavior would be available to any instance of such a schema).  The property would be named something like `relative_location` and would consist of a **Unix**-style path value (alphanumeric, forward-slash, hyphens, and underscores allowed).   The UI and UX may very well "marry" `relative_location` with `name` in the form `relative_location/name` but the request to the server would separate them out to their respective properties.

When a value for `relative_location` is present, the metadata service conveys that value to the configured metadata store, and only _it_ interprets how it should be deciphered.  For example, the `FileMetadataStore` would interpret `relative_location` as the set of directories _within the found namespace location_ in which to store the instance.  A `NoSQLMetadataStore` would use `relative_location` however it needs - depending on the organization of the records.  While a `GitMetadataStore` might interpret the first portion of `relative_location` as the Git organization, the second as the repo, and any remaining "portions" as sub-directories within the repo.  In this sense, the _transient_ property _is_ persisted, but not as an element of the JSON instance itself.  Actually, it would be treated much like another transient property - `name`.

> In the future, we could look into enhancing the git extension to point it to directories outside of the jupyter working directory to enable git operations with code-snippets.

I don't think this will be possible unless the _Namespace location_ is also outside of the jupyter working directory.  That is, **all metadata instances must reside within their respective namespace locations for discovery**.  However, were we to promote `Namespace` to full object citizenship, then, besides the 4 _system locations_ we could also configure alternate locations.  It's just that the namespace instance itself could only reside in the various 4 system locations since it must be deterministically discoverable.  ((I think we'd probably just want to support one location for namespace instances because I believe we'd also want a sub-directory of _that_ location to house the schemas relative to the namespace and its that `schemas` directory in which custom (bring your own schemas) - as well as system schemas - would be located.))

**Comment by jondoering** [comment]
@lresende 
Thanks for the response - the points you mention would pretty much fulfill what I had in mind and the use case I tried to solve. I think, having a way to store new code snipped automatically in the respective git folder would be very helpful. I very much like @kevin-bates idea of having that as another property.
